future_test.go

```diff
@@ -0,0 +1,235 @@
+	} else if it.Value().Code != code {
+		t.Errorf("An unexpected response code %d, expected %d", it.Value().Code, code)
+	}
+
+	if it.Value() != resp {
+		t.Errorf("An unexpected response %v, expected %v", it.Value(), resp)
+	}
+}
+
+func assertResponseIteratorFinished(t testing.TB, it ResponseIterator) {
+	t.Helper()
+
+	if it.Err() != nil {
+		t.Errorf("An unexpected iteration error: %q", it.Err().Error())
+	}
+	if it.Value() != nil {
+		t.Errorf("An unexpected value %v", it.Value())
+	}
+}
+
+func TestFutureGetIteratorNoItems(t *testing.T) {
+	fut := NewFuture()
+
+	it := fut.GetIterator()
+	if it.Next() {
+		t.Errorf("An unexpected next value.")
+	} else {
+		assertResponseIteratorFinished(t, it)
+	}
+}
+
+func TestFutureGetIteratorNoResponse(t *testing.T) {
+	push := &Response{}
+	fut := NewFuture()
+	fut.AppendPush(push)
+
+	if it := fut.GetIterator(); it.Next() {
+		assertResponseIteratorValue(t, it, PushCode, push)
+		if it.Next() == true {
+			t.Errorf("An unexpected next value.")
+		}
+		assertResponseIteratorFinished(t, it)
+	} else {
+		t.Errorf("A push message expected.")
+	}
+}
+
+func TestFutureGetIteratorNoResponseTimeout(t *testing.T) {
+	push := &Response{}
+	fut := NewFuture()
+	fut.AppendPush(push)
+
+	if it := fut.GetIterator().WithTimeout(1 * time.Nanosecond); it.Next() {
+		assertResponseIteratorValue(t, it, PushCode, push)
+		if it.Next() == true {
+			t.Errorf("An unexpected next value.")
+		}
+		assertResponseIteratorFinished(t, it)
+	} else {
+		t.Errorf("A push message expected.")
+	}
+}
+
+func TestFutureGetIteratorResponseOnTimeout(t *testing.T) {
+	push := &Response{}
+	resp := &Response{}
+	fut := NewFuture()
+	fut.AppendPush(push)
+
+	var done sync.WaitGroup
+	var wait sync.WaitGroup
+	wait.Add(1)
+	done.Add(1)
+
+	go func() {
+		defer done.Done()
+
+		var it ResponseIterator
+		var cnt = 0
+		for it = fut.GetIterator().WithTimeout(5 * time.Second); it.Next(); {
+			code := PushCode
+			r := push
+			if cnt == 1 {
+				code = OkCode
+				r = resp
+			}
+			assertResponseIteratorValue(t, it, code, r)
+			cnt += 1
+			if cnt == 1 {
+				wait.Done()
+			}
+		}
+		assertResponseIteratorFinished(t, it)
+
+		if cnt != 2 {
+			t.Errorf("An unexpected count of responses %d != %d", cnt, 2)
+		}
+	}()
+
+	wait.Wait()
+	fut.SetResponse(resp)
+	done.Wait()
+}
+
+func TestFutureGetIteratorFirstResponse(t *testing.T) {
+	resp1 := &Response{}
+	resp2 := &Response{}
+	fut := NewFuture()
+	fut.SetResponse(resp1)
+	fut.SetResponse(resp2)
+
+	if it := fut.GetIterator(); it.Next() {
+		assertResponseIteratorValue(t, it, OkCode, resp1)
+		if it.Next() == true {
+			t.Errorf("An unexpected next value.")
+		}
+		assertResponseIteratorFinished(t, it)
+	} else {
+		t.Errorf("A response expected.")
+	}
+}
+
+func TestFutureGetIteratorFirstError(t *testing.T) {
+	const errMsg1 = "error1"
+	const errMsg2 = "error2"
+
+	fut := NewFuture()
+	fut.SetError(errors.New(errMsg1))
+	fut.SetError(errors.New(errMsg2))
```

Comment:
Is this situation possible in real life? [2]

Reply:
Not now. But this test determines the behavior of Future for the case. So I think it's ok. [2]